subsys/bluetooth/audio/vocs_client.c

```diff
@@ -0,0 +1,668 @@
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/vocs.h>
+
+#include "vocs_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_VOCS_CLIENT)
+#define LOG_MODULE_NAME bt_vocs_client
+#include "common/log.h"
+
+static struct bt_vocs vocs_insts[CONFIG_BT_MAX_CONN * CONFIG_BT_VOCS_CLIENT_MAX_INSTANCE_COUNT];
+
+static struct bt_vocs *lookup_vocs_by_handle(struct bt_conn *conn, uint16_t handle)
+{
+	__ASSERT(handle != 0, "Handle cannot be 0");
+	__ASSERT(conn, "Conn cannot be NULL");
+
+	for (int i = 0; i < ARRAY_SIZE(vocs_insts); i++) {
+		if (vocs_insts[i].cli.conn == conn &&
+		    vocs_insts[i].cli.active &&
+		    vocs_insts[i].cli.start_handle <= handle &&
+		    vocs_insts[i].cli.end_handle >= handle) {
+			return &vocs_insts[i];
+		}
+	}
+	BT_DBG("Could not find VOCS instance with handle 0x%04x", handle);
+	return NULL;
+}
+
+uint8_t vocs_client_notify_handler(struct bt_conn *conn, struct bt_gatt_subscribe_params *params,
+				   const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, handle);
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (data) {
+		if (handle == inst->cli.state_handle) {
+			if (length == sizeof(inst->cli.state)) {
+				memcpy(&inst->cli.state, data, length);
+				BT_DBG("Inst %p: Offset %d, counter %u",
+				       inst,
+				       inst->cli.state.offset,
+				       inst->cli.state.change_counter);
+				if (inst->cli.cb && inst->cli.cb->state) {
+					inst->cli.cb->state(conn, inst, 0, inst->cli.state.offset);
+				}
+			}
+		} else if (handle == inst->cli.desc_handle) {
+			char desc[MIN(CONFIG_BT_L2CAP_RX_MTU, BT_ATT_MAX_ATTRIBUTE_LEN) + 1];
+
+			/* Truncate if too large */
+			length = MIN(sizeof(desc) - 1, length);
+
+			memcpy(desc, data, length);
+			desc[length] = '\0';
+			BT_DBG("Inst %p: Output description: %s", inst, log_strdup(desc));
+			if (inst->cli.cb &&
+			    inst->cli.cb->description) {
+				inst->cli.cb->description(conn, inst, 0, desc);
+			}
+		} else if (handle == inst->cli.location_handle) {
+			if (length == sizeof(inst->cli.location)) {
+				memcpy(&inst->cli.location, data, length);
+				BT_DBG("Inst %p: Location %u", inst, inst->cli.location);
+				if (inst->cli.cb &&
+				    inst->cli.cb->location) {
+					inst->cli.cb->location(conn, inst, 0, inst->cli.location);
+				}
+			}
+		}
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t vocs_client_read_offset_state_cb(
+	struct bt_conn *conn, uint8_t err, struct bt_gatt_read_params *params,
+	const void *data, uint16_t length)
+{
+	uint8_t cb_err = err;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, params->single.handle);
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+	inst->cli.busy = false;
+
+	if (data) {
```

Comment:
See https://github.com/zephyrproject-rtos/zephyr/pull/31893#discussion_r597681691

Reply:
Actually, since this is for a read: If the client reads a characteristic, and the value returned by the GATT server has length 0, the data pointer will be NULL (and length as well). It's perfectly valid for the server to return 0 bytes from a read.